python/docs/source/user_guide/sql/python_udtf.rst

```diff
@@ -0,0 +1,222 @@
+..  Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+..    http://www.apache.org/licenses/LICENSE-2.0
+
+..  Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+===========================================
+Python User-defined Table Functions (UDTFs)
+===========================================
+
+Spark 3.5 introduces Python user-defined table functions (UDTFs), a new type of user-defined function. 
+Unlike scalar functions that return a single result value, a UDTF is invoked in the FROM clause and returns 
+an entire relation as output. Each UDTF call can accept zero or more arguments. 
+These arguments can be scalar constant expressions or separate input relations.
+
+Implementing a Python UDTF
+--------------------------
+
+.. currentmodule:: pyspark.sql.functions
+
+To implement a Python UDTF, you can define a class implementing the methods:
+
+.. code-block:: python
+
+    class PythonUDTF:
+
+        def __init__(self) -> None:
+            """
+            Initialize the user-defined table function (UDTF).
+
+            This method serves as the default constructor and is called once when the
+            UDTF is instantiated on the executor side.
+            
+            Any class fields assigned in this method will be available for subsequent
+            calls to the `eval` and `terminate` methods.
+
+            Notes
+            -----
+            - This method does not accept any extra arguments.
+            - You cannot create or reference the Spark session within the UDTF. Any
+              attempt to do so will result in a serialization error.
+            """
+            ...
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """
+            Evaluate the function using the given input arguments.
+
+            This method is required and must be implemented.
+            
+            The arguments provided to the UDTF call are mapped to the values in the
+            `*args` list sequentially. Each provided scalar expression maps to exactly
+            one value in this `*args` list. Each provided TABLE argument of N columns
```

Comment:
@dtenedor Here's the line

Reply:
@allisonwang-db it turns out this part about `TABLE` arguments is wrong (I think I suggested it before, sorry). Instead of:

```
Each provided TABLE argument of N columns
maps to exactly N values in this `*args` list, in the order of the columns
as they appear in the table.
```

it should be something like

```
Each provided TABLE argument maps to a pyspark.sql.Row object containing
the columns in the order they appear in the provided input relation.
```

Reply:
Updated!